text/3621-extended_hrtb.md

```diff
@@ -0,0 +1,93 @@
+- Feature Name: `extended_hrtb`
+- Start Date: 2022-05-08
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/3621)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This feature add a way to bound universal quantification of HTRBs to allow more APIs.
+
+# Motivation
+[motivation]: #motivation
+
+The HRTB construct has allowed us to talk about closures that are generic over lifetimes. However, initially this has not included a way to restrict these generic lifetimes, thus, in practice, it turned out to be useful in only a handful of scenarios.
+
+The goal of the proposal is to allow bounded universal quantification over lifetimes.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The new syntax looks like this:\
+`for<'a where 'b: 'a>` - it's called extended HRTB\
+This construct reads as "for all lifetimes that `'b` outlives".
+
+This form of HRTB allow you to make the following interface:\
+(The example is taken from Sabrina's post, see Prior Art)
+```rust
+fn print_items<'i,I>(mut iter: I)
+where
+	I: LendingIterator + 'i,
+	for<'a where 'i: 'a> I::Item<'a>: Debug,
```

Comment:
Here's something that is a bit confusing to me: shouldn't the compiler be able to infer `'i: 'a` based upon the fact that `I: 'i`? I get that it doesn't now, but I'm wondering whether this would still be useful if that were the case.

Reply:
Won't it change the meaning of the entire construct? And if so, how do we write the old meaning?

Reply:
I don't see how it would change the meaning of the construct; if `I: 'i` then isn't every lifetime inside of `I` bound by `'i`? Or am I massively misunderstanding something?

Reply:
> then isn't every lifetime inside of `I` bound by `'i`?

This is not necessary true: given a GAT one can construct a type with any parameters, these may easily outlive the `Self` type if not said otherwise.

Reply:
I think I'm probably just misremembering various extensions that have been made to lifetimes since 1.0, alongside some of the newer features like GAT in the queue. I'll try and remember to read the linked article and see if I can wrap my head around it better.

Thanks for trying to help explain things.